### ISSUE TITLE
[BUG] Variable substitution for `dotnet` task type (e.g. `${workspaceFolder}`) regressed in current builds

### ISSUE THREAD

**Issue by giddingsra0** [microsoft/vscode-dotnettools#803]
### Describe the Issue

In a multi-project solution that are all nested within the same root project directory (see the Steps To Reproduce section for example), a build task referencing a specific project `.csproj` file will fail to find the file unless the full file path is used in the `"file"` attribute of the task definition (i.e. `/Users/{username}/some/project/path`).

The task type used is `dotnet`. See the Steps To Reproduce section for example.

I have tried the following task definition variations to see what works and what doesn't:

1.
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "dotnet",
      "task": "build",
      "problemMatcher": [
        "$msCompile"
      ],
      "group": "build",
      "label": "dotnet: build",
      "file": "${workspaceFolder}/project1/project1.csproj"
    }
  ]
}
```

Output, does not resolve the `${workspaceFolder}` variable:

```bash
 *  Executing task: dotnet: build

dotnet build ${workspaceFolder}/project1/project1.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary 
MSBuild version 17.4.1+9a89d02ff for .NET
MSBUILD : error MSB1009: Project file does not exist.
Switch: ${workspaceFolder}/project1/project1.csproj

 *  The terminal process failed to launch (exit code: 1). 
 *  Terminal will be reused by tasks, press any key to close it. 
```

2.
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "dotnet",
      "task": "build",
      "options": {
        "cwd": "${workspaceFolder}"
      },
      "problemMatcher": [
        "$msCompile"
      ],
      "group": "build",
      "label": "dotnet: build",
      "file": "project1/project1.csproj"
    }
  ]
}
```

Output, cannot find the project file:

```bash
 *  Executing task: dotnet: build

dotnet build project1/project1.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary 
MSBuild version 17.4.1+9a89d02ff for .NET
MSBUILD : error MSB1009: Project file does not exist.
Switch: project1/project1.csproj

 *  The terminal process failed to launch (exit code: 1). 
 *  Terminal will be reused by tasks, press any key to close it. 
```

3.
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "dotnet",
      "task": "build",
      "options": {
        "cwd": "${workspaceFolder}"
      },
      "problemMatcher": [
        "$msCompile"
      ],
      "group": "build",
      "label": "dotnet: build",
      "file": "${cwd}/project1/project1.csproj"
    }
  ]
}
```

Output, does not resolve the `${cwd}` variable:

```bash
 *  Executing task: dotnet: build

dotnet build ${cwd}/project1/project1.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary 
MSBuild version 17.4.1+9a89d02ff for .NET
MSBUILD : error MSB1009: Project file does not exist.
Switch: ${cwd}/project1/project1.csproj

 *  The terminal process failed to launch (exit code: 1). 
 *  Terminal will be reused by tasks, press any key to close it. 
```

4 and 5. Same as above but using `${workspaceRoot}` instead of `${workspaceFolder}`

6.
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "dotnet",
      "task": "build",
      "problemMatcher": [
        "$msCompile"
      ],
      "group": "build",
      "label": "dotnet: build",
      "file": "./project1/project1.csproj"
    }
  ]
}
```

Output, cannot find project file:

```bash
 *  Executing task: dotnet: build

dotnet build ./project1/project1.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary 
MSBuild version 17.4.1+9a89d02ff for .NET
MSBUILD : error MSB1009: Project file does not exist.
Switch: ./project1/project1.csproj

 *  The terminal process failed to launch (exit code: 1). 
 *  Terminal will be reused by tasks, press any key to close it. 
```

7.
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "dotnet",
      "task": "build",
      "problemMatcher": [
        "$msCompile"
      ],
      "group": "build",
      "label": "dotnet: build",
      "file": "/Users/{username}/some/project/path/project1/project1.csproj"
    }
  ]
}
```

Output, builds successfully:

```bash
 *  Executing task: dotnet: build

dotnet build /Users/{username}/some/project/path/project1/project1.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary 
MSBuild version 17.4.1+9a89d02ff for .NET
Determining projects to restore...
  All projects are up-to-date for restore.
  project1 -> /Users/{username}/some/project/path/project1/bin/Debug/net7.0/project1.dll
  ... successfully builds all project dependencies as well
 *  Terminal will be reused by tasks, press any key to close it. 
```

### Steps To Reproduce

1. Install the C# Dev Kit extension in VS Code at version 1.1.14 or newer (this issue is not present in prerelease version 1.1.12)
2. Create a new solution
3. Add two or more projects to the solution, at least one that is runnable, in the following directory structure:
```bash
|
|---solutionRoot
     |
     |--- .vscode
     |      |
     |      |--- launch.json
     |      |--- tasks.json
     |
     |--- project1
     |     |
     |     |--- Program.cs
     |     |--- project1.csproj
     |
     |--- project2
     |     |
     |     |--- Program.cs
     |     |--- project2.csproj
     |
     |--- solutionRoot.sln
```
4. Add a VS Code launch configuration in `launch.json` for one of the applications with an build task in `tasks.json` that looks like the following:
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "dotnet",
      "task": "build",
      "problemMatcher": [
        "$msCompile"
      ],
      "group": "build",
      "label": "dotnet: build",
      "file": "${workspaceFolder}/project1/project1.csproj"
    }
  ]
}
```
5. Add the newly created build task as a `preLaunchTask` for the launch configuration in `launch.json`.
6. Try to run the launch configuration and watch the output report the following:
```bash
 *  Executing task: dotnet: build

dotnet build ${workspaceFolder}/project1/project1.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary 
MSBuild version 17.4.1+9a89d02ff for .NET
MSBUILD : error MSB1009: Project file does not exist.
Switch: ${workspaceFolder}/project1/project1.csproj

 *  The terminal process failed to launch (exit code: 1). 
 *  Terminal will be reused by tasks, press any key to close it. 
```
7. Downgrade to version 1.1.12 or older of the C# Dev Kit extension in VS Code and reload the window
8. Try to run the launch configuration and watch the build task resolve the `${workspaceFolder}` to the path of the current workspace and execute successfully.

### Expected Behavior

I would expect the following:
- `${workspaceFolder}` to resolve correctly to the current location of the workspace root directory on the file system
- the build task to successfully build the project and all project dependencies

### Environment Information

- OS: MacOS Sonoma 14.2 (23C64) with Apple Silicon
- VS Code version: 
```
Version: 1.85.0 (Universal)
Commit: af28b32d7e553898b2a91af498b1fb666fdebe0c
Date: 2023-12-06T18:18:04.614Z (6 days ago)
Electron: 25.9.7
ElectronBuildId: 25551756
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Darwin arm64 23.2.0
```
- C# Dev Kit Extension Version: issue is present on version 1.1.14 and newer
- .NET version: 7.0.102

**Comment by dotMorten** [comment]
I'm seeing this and spent a lot of time trying to get it to work - I've given up for now.

**Comment by giddingsra0** [comment]
> I'm seeing this and spent a lot of time trying to get it to work - I've given up for now.

Glad it isn't just me!  😆  For now, I have just downgraded to the v1.1.12 pre-release version and it is working as expected. I am planning on staying on that version until this is fixed.

**Comment by timheuer** [comment]
What you want is type=`process` and use the `args` (the 'generate build tasks' command in VS Code would help you generate some of these).  The result is this:

```
{
            "label": "build",
            "command": "dotnet",
            "type": "process",
            "args": [
                "build",
                "${workspaceFolder}/feedback-reorder/feedback-reorder.csproj",
                "/property:GenerateFullPaths=true",
                "/consoleloggerparameters:NoSummary;ForceNoAlign"
            ],
            "problemMatcher": "$msCompile"
        }
```

Were you seeing the type:dotnet used elsewhere with that pattern? there is a launch type for dotnet, but not a custom task type if I recall other than built-in tasks (which are build/rebuild/publish).

I assume you wanted to customize different build tasks with multi-projects, so using the `process` type is the pattern you want here and `args` in VSCode are replaced but only in certain situations as defined by VSCode: https://code.visualstudio.com/docs/editor/tasks#_variable-substitution

**Comment by giddingsra0** [comment]
Okay, I have no problem going back to using this pattern like I was before C# Dev Kit was released. This does raise a couple questions though:
- If using custom tasks `"type": "dotnet"` is not a valid use case, why did this work as expected in v1.1.12 of C# Dev Kit but not in v1.1.14 or later? What changed?
- I am still able to customize the built-in tasks (build/rebuild/publish) that have `"type": "dotnet"` and add new tasks of `"type": "dotnet"` (which VS Code task autocompletion also provides valid options for) currently. Why is this allowed if the configurations are not respected?

**Comment by timheuer** [comment]
The configurations are respected, but I'm just pointing out that VSCode does not do replacement in that task execution engine (for lack of a better word) in areas other than documented in variable-substitution.  

You saw variable substitution in that task config working prior? Trying now and will report back on that version.  I don't think we (the extension) are involved in doing the arg/variable substitution but instead just get what is passed.

**Comment by timheuer** [comment]
@WardenGnaw do you own the custom task?

**Comment by giddingsra0** [comment]
@timheuer I appreciate your team's and your time on this!

**Comment by giddingsra0** [comment]
Any updates? Just checking in on this to see where it landed.

**Comment by timheuer** [comment]
> Any updates? Just checking in on this to see where it landed.

We're in a bit of 'everyone out on vacation' mode here unfortunately.  Will need @WardenGnaw to look at this when back

**Comment by giddingsra0** [comment]
Okay! No worries. I was just checking in! Thank you

**Comment by giddingsra0** [comment]
Hi there! Just wanted to check in on this since it has been about a month since I last checked on this.

**Comment by giddingsra0** [comment]
Hi there @timheuer and @WardenGnaw , just checking in on this to get a status update.

**Comment by alexisbg** [comment]
Wouldn't it be simpler to allow relative paths?
```json
"file": "project1/project1.csproj"
```
or
```json
"file": "./project1/project1.csproj"
```

**Comment by giddingsra0** [comment]
> Wouldn't it be simpler to allow relative paths?

I would also be totally fine with this. Basically, all I need is to be able to share launch configs with the rest of my team without having to specify the full path on the file system.

**Comment by WardenGnaw** [comment]
The issue seems to not be with `launch.json` with `type: "dotnet"`. The issue is the TaskProvider for `dotnet` that @adrianvmsft worked on.

**Comment by giddingsra0** [comment]
@adrianvmsft Hi there! I just wanted to check on the status of this issue and maybe an estimated timeline when this will be resolved?

**Comment by adrianvmsft** [comment]
There was indeed an accidental regression where variable substitution stopped working for the build tasks generated by the C# Dev Kit extension (`type: dotnet`).
Thank you for narrowing down the builds where the regression was introduced - that was very helpful in identifying the change that caused it (it was a change where the task execution was changed to custom to enable error deduplication between intellisense and build which needed some special handling for variable substitution).
I submitted a fix for this regression.
